### PR TITLE
Remove rollback timeout

### PR DESCRIPTION
**Pull request by geof90**
Breaking Changes:
1. The notion of a rollback timeout is now removed. In the event that an app crashes after an update, CodePush will automatically rollback and load the previous version when the user starts the app again.
2. If you are using the advanced API, you MUST call notifyApplicationReady() somewhere in your bundle. This is to notify CodePush that the update is okay. Otherwise, CodePush will treat the update as a faulty one and rollback when the app next restarts. If you are using the Sync API, notifyApplicationReady() is called inside sync(), so you do not need to change anything.

Bug fix:
1. On Android, the updates were not being loaded when the app is in developer mode, even when the dev server is not running. This was caused by React Native caching the bundle downloaded from the dev server, and since dev server bundles always take precedence over any other locally stored bundles, CodePush updates will never be run even after they are installed. This fixes that issue by simply deleting the cached file when an update is applied.

### CODE REVIEW

File: android/app/src/main/java/com/microsoft/codepush/react/CodePush.java [hunk 1]
```diff
@@ -204,52 +187,60 @@ private void saveFailedUpdate(String packageHash) {
...
+        if (pendingUpdateString == null) {
+            return null;
+        }
+
+        try {
+            JSONObject pendingUpdate = new JSONObject(pendingUpdateString);
+            return pendingUpdate;
+        } catch (JSONException e) {
+            // Should not happen.
+            throw new CodePushUnknownException("Unable to parse pending update metadata " +
```
Comment (dlebu) [inline android/app/src/main/java/com/microsoft/codepush/react/CodePush.java]: How about handling the exception silently and not re-throwing here?

**Comment by dlebu** [comment]
Do we have any automated tests for this change?

**Comment by geof90** [comment]
@dlebu Not yet. We are behind on implementing tests for a few features, e.g the sync implementation currently has no automated test coverage. There is a task in the backlog assigned to me for writing all automated tests. We just want to quickly get this out there because right now a number of customers have given us feedback that the present way that rollbacks work is not working for them (app crashes before notifyApplicationReady is called that terminates the app preventing the rollback timer from ever kicking in).

File: CodePush.m [hunk 2]
```diff
@@ -130,16 +117,17 @@ - (void)initializeUpdateAfterRestart
...
     if (pendingUpdate) {
         _isFirstRunAfterUpdate = YES;
-        int rollbackTimeout = [pendingUpdate[PendingUpdateRollbackTimeoutKey] intValue];
-        if (0 != rollbackTimeout) {
-            dispatch_async(dispatch_get_main_queue(), ^{
-                [self startRollbackTimer:rollbackTimeout];
-            });
+        BOOL wasInitialized = [pendingUpdate[PendingUpdateWasInitializedKey] boolValue];
+        if (wasInitialized) {
+            // Pending update was initialized, but notifiyApplicationReady was not called.
```
Comment (dtivel) [inline CodePush.m]: notifyApplicationReady

File: CodePush.m [hunk 3]
```diff
@@ -216,38 +203,35 @@ - (void)saveFailedUpdate:(NSString *)packageHash
 }
 
 /*
+ * This method is called in notifyApplicationReady to register the fact that
+ * the pending update succeeded and therefore can be removed.
```
Comment (dtivel) [inline CodePush.m]: I don't think you should include caller information in comments.  Just state what this method does.

File: CodePush.m [hunk 4]
```diff
@@ -130,16 +117,17 @@ - (void)initializeUpdateAfterRestart
     
     if (pendingUpdate) {
         _isFirstRunAfterUpdate = YES;
-        int rollbackTimeout = [pendingUpdate[PendingUpdateRollbackTimeoutKey] intValue];
-        if (0 != rollbackTimeout) {
-            dispatch_async(dispatch_get_main_queue(), ^{
-                [self startRollbackTimer:rollbackTimeout];
-            });
+        BOOL wasInitialized = [pendingUpdate[PendingUpdateWasInitializedKey] boolValue];
+        if (wasInitialized) {
```
Comment (dtivel) [inline CodePush.m]: Not crazy about this name.  I think a better name would be "isUpdateLoading."  From the point that the bridge reloads to now, that is what's happening.  If the value is true here, then the update loaded successfully.

File: README.md [hunk 5]
```diff
@@ -248,9 +248,9 @@ When you require `react-native-code-push`, the module object provides the follow
 
 * [getCurrentPackage](#codepushgetcurrentpackage): Retrieves the metadata about the currently installed update (e.g. description, installation time, size).
 
-* [notifyApplicationReady](#codepushnotifyapplicationready): Notifies the CodePush runtime that an installed update is considered successful. This is an optional API, but is useful when you want to expicitly enable "rollback protection" in the event that an exception occurs in code that you've deployed to production. 
+* [notifyApplicationReady](#codepushnotifyapplicationready): Notifies the CodePush runtime that an installed update is considered successful. If you are manually checking for and install updates (i.e. not using the [sync](#codepushsync) method to handle it all for you), then this method **MUST** be called, otherwise CodePush will regard the update as a failed one and rollback to the previous version when the app next restarts.
```
Comment (dtivel) [inline README.md]: "...and installing updates..."

"...be called; otherwise, CodePush will treat the update as failed and..."

File: README.md [hunk 6]
```diff
@@ -248,9 +248,9 @@ When you require `react-native-code-push`, the module object provides the follow
...
 
-* [restartApp](#codepushrestartapp): Immediately restarts the app. If there is an update pending, it will be immediately displayed to the end-user, and the rollback timer (if specified when installing the update) will begin. Otherwise, calling this method simply has the same behavior as the end-user killing and restarting the process.
+* [restartApp](#codepushrestartapp): Immediately restarts the app. If there is an update pending, it will be immediately displayed to the end-user. Otherwise, calling this method simply has the same behavior as the end-user killing and restarting the process.
```
Comment (dtivel) [inline README.md]: "end user" (no hyphen)

File: README.md [hunk 7]
```diff
@@ -310,9 +310,7 @@ codePush.getCurrentPackage()
...
-
-If the `rollbackTimeout` parameter was not specified, the CodePush runtime will not enforce any automatic rollback behavior, and therefore, calling this function is not required and will result in a no-op.
+Notifies the CodePush runtime that a freshly installed update should be considered successful, and therefore, an automatic client-side rollback isn't necessary. It is mandatory to call this function somewhere in the code of the updated bundle. Otherwise, when the app next restarts, the CodePush runtime will assume that the installed update has failed and roll back to the previous version. This behavior exists to help ensure that your end-users aren't blocked by a broken update.
```
Comment (dtivel) [inline README.md]: "end users" (not hyphenated)

File: README.md [hunk 8]
```diff
@@ -459,11 +455,7 @@ Contains details about an update that has been downloaded locally or already ins
...
-<br /><br />
-*Note: The "rollback timer" doesn't start until the update has actually become active. If the `installMode` is `IMMEDIATE`, then the rollback timer will also start immediately. However, if the `installMode` is `UPDATE_ON_RESTART` or `UPDATE_ON_RESUME`, then the rollback timer will start the next time the app starts or resumes, not at the point that you called `install`.*
+- __install(installMode: CodePush.InstallMode = CodePush.InstallMode.ON_NEXT_RESTART): Promise&lt;void&gt;__: Installs the update by saving it to the location on disk where the runtime expects to find the latest version of the app. The `installMode` parameter controls when the changes are actually presented to the end-user. The default value is to wait until the next app restart to display the changes, but you can refer to the [`InstallMode`](#installmode) enum reference for a description of the available options and what they do.
```
Comment (dtivel) [inline README.md]: "end user" (no hyphen)

**Comment by dtivel** [comment]
Looks good.

**Pull request merged**